### ISSUE TITLE
Listing is returing 503 upon upgrade

### ISSUE THREAD

**Issue by rvadim** [minio/minio#12586]
Unable to cp/stat/head some objects after upgrade.

## Expected Behavior
No currupted files.

## Current Behavior
After upgrade from `2020-04-15T19-42-18Z` to `2021-06-07T21-40-51Z` and then to `RELEASE.2021-06-17T00-10-46Z` we have:
```
$ mc stat my-minio/bucket/report.txt
mc: <ERROR> Unable to stat `my-minio/bucket/report.txt`. 503 Service Unavailable.
```
There are a lot of messages at the log like:
```
API: SYSTEM()
Time: 07:17:00 UTC 06/28/2021
DeploymentID: 776475dd-8193-4864-a757-09f8af1bd12d
Error: Storage resources are insufficient for the read operation .minio.sys/buckets/bucket/.metacache/index.s2 (cmd.InsufficientReadQuorum)
       1: cmd/metacache-manager.go:84:cmd.(*metacacheManager).initManager.func1()
```
`Heal` reports gray status for 10% of objects.

Also I cannot delete such objects:
```
$ mc rm my-minio/bucket/animated.gif
mc: <ERROR> Failed to remove `my-minio/bucket/animated.gif`. 503 Service Unavailable
```

## Possible Solution
No idea right now.

## Steps to Reproduce (for bugs)
1. Setup distributed minio 2020-04-15T19-42-18Z` 
2. Fill up with ~1M objects
3. Upgrade to `2020-04-15T19-42-18Z` to `2021-06-07T21-40-51Z` and then to `RELEASE.2021-06-17T00-10-46Z`
4. Check all files (mc mirror)

## Context
We need to heal or delete files. Can you suggest workaround for deletion?

## Regression
-

## Your Environment
* Our cluster has 8 nodes, 32 disks, 100TB of data, 3.7M objects.
* RELEASE.2021-06-17T00-10-46Z
* Kubernetes custom setup with 2 zones. [archive with yamls: zones.zip](https://github.com/minio/minio/files/6692073/zones.zip)
* Ubuntu 20.04

**Comment by harshavardhana** [comment]
> We need to heal or delete files. Can you suggest workaround for deletion?

@rvadim from the error that you are reporting it looks like Listing is returning such errors, can you provide `mc rm --debug` output for this?  you can also purge 

> rm -rf /backend_drive/.minio.sys/buckets/*/.metacache/

and then `mc admin service restart` your deployment.

**Comment by rvadim** [comment]
mc rm --debug log:
[debug.log](https://github.com/minio/minio/files/6724383/debug.log)

**Comment by klauspost** [comment]
Seems like some drives/servers are offline. Attach `mc admin info --json myminio >info.json` as well.

**Comment by rvadim** [comment]
[info-pretty.json.gz](https://github.com/minio/minio/files/6724757/info-pretty.json.gz)

**Comment by harshavardhana** [comment]
@rvadim are you sure you didn't upgrade to any releases?  also please attach the `animated.gif` file from all the servers here? for investigation?

```
RELEASE.2021-05-11T23-27-41Z
RELEASE.2021-05-16T05-32-34Z
RELEASE.2021-05-18T00-53-28Z
RELEASE.2021-05-20T22-31-44Z
RELEASE.2021-05-22T02-34-39Z
RELEASE.2021-05-26T00-22-46Z
RELEASE.2021-05-27T22-06-31Z
```

in between?

**Comment by rvadim** [comment]
I tring to prepare archives from all disks, but it is not so easy.

As I wrote, we upgraded directly from `2020-04-15T19-42-18Z` to `2021-06-07T21-40-51Z`.

**Comment by rvadim** [comment]
Can you please suggest any steps to remove the currupted files?

**Comment by harshavardhana** [comment]
@rvadim it would be `mc admin heal -r --remove`  this will attempt to remove unrecoverable objects, but it is important that we see these objects to ensure something odd is not causing these to simply corrupt.

**Comment by rvadim** [comment]
Sorry, all currupted files are under NDA. I can share one of it with you on private email with NDA if you want.

About heal:
```
minio@minio-mc-656fd9d46d-6k59g:~$ mc admin heal -r --remove my-minio/buildserver-test
 ◑  buildserver-test/SelfTes…rtifacts/67290416/artifacts/972.txt
    0/4 objects; 109 B in 2s
    ┌────────┬─────┬─────────────────────┐
    │ Green  │ 113 │  96.6% ████████████ │
    │ Yellow │   0 │   0.0%              │
    │ Red    │   0 │   0.0%              │
    │ Grey   │   4 │   3.4% █            │
    └────────┴─────┴─────────────────────┘
```
Repeated launches give the same result.
```
mc rm -r --force my-minio/buildserver-test
Removing `....
mc: <ERROR> Failed to remove `space/buildserver-test` recursively. Resource requested is unreadable, please reduce your request rate
```
Looks like the `--remove` is not working(

**Comment by harshavardhana** [comment]
> Sorry, all currupted files are under NDA. I can share one of them with you on private email with NDA if you want.

Unfortunately, we can't do that without a formal relationship; feel free to subscribe to our support network https://min.io/pricing - then everything can be formalized.

**Comment by rvadim** [comment]
Looks like removing directory by object name from all disks works:
```
rm -Rf /data-*/bucket/report.txt
```
Heal don't show any gray objects after deletion and I can upload file to minio from backup.

**Comment by harshavardhana** [comment]
> Looks like removing directory by object name from all disks works:
> 
> ```
> rm -Rf /data-*/bucket/report.txt
> ```
> 
> Heal don't show any gray objects after deletion and I can upload file to minio from backup.

Okay 👌

**Comment by haslersn** [comment]
I experienced a 503 like this after upgrade, too. The problem was probably already present before the upgrade, but the symptom was different before the upgrade. More specifically:

1. `RELEASE.2021-04-06T23-11-00Z` (i.e., before upgrade):
   - `mc ls <object>` sometimes lists it and sometimes doesn't.
   - `mc cat <object>` sometimes returns `Object does not exist.` and sometimes returns `The specified key does not exist.`.
2. `RELEASE.2021-06-17T00-10-46Z` was the only intermediate version I deployed.
3. `RELEASE.2021-07-15T22-27-34Z` (i.e., after upgrade):
   - `mc ls <object>` sometimes lists it and sometimes doesn't. (Same as before.)
   - `mc cat <object>` always returns `503 Service Unavailable.`
   - `mc stat <object>` always returns `503 Service Unavailable.`
   - `mc admin heal <object>` shows 24 green, 0 yellow/read/grey, but the problem persists. (Apparently the heal command also checks objects in `.minio.sys`, that's why there are 24 green instead of 1.)

## More information

I have 3 servers and 2 pools (zone-0 and zone-1). Each pool has 4 drives per server. On each drive that belongs to the second pool (i.e., zone-1) there are the following files present:

- `<object>/xl.meta`
- `<object>/0dcf7fc5-d7b6-4895-95f2-33da7c09189d/part.1`

So the object doesn't seem to be missing on any drive.

**Comment by harshavardhana** [comment]
> I experienced a 503 like this after upgrade, too. The problem was probably already present before the upgrade, but the symptom was different before the upgrade. More specifically:
> 
> 1. `RELEASE.2021-04-06T23-11-00Z` (i.e., before upgrade):
>    
>    * `mc ls <object>` sometimes lists it and sometimes doesn't.
>    * `mc cat <object>` sometimes returns `Object does not exist.` and sometimes returns `The specified key does not exist.`.
> 2. `RELEASE.2021-06-17T00-10-46Z` was the only intermediate version I deployed.
> 3. `RELEASE.2021-07-15T22-27-34Z` (i.e., after upgrade):
>    
>    * `mc ls <object>` sometimes lists it and sometimes doesn't. (Same as before.)
>    * `mc cat <object>` always returns `503 Service Unavailable.`
>    * `mc stat <object>` always returns `503 Service Unavailable.`
>    * `mc admin heal <object>` shows 24 green, 0 yellow/read/grey, but the problem persists. (Apparently the heal command also checks objects in `.minio.sys`, that's why there are 24 green instead of 1.)
> 
> ## More information
> I have 3 servers and 2 pools (zone-0 and zone-1). Each pool has 4 drives per server. On each drive that belongs to the second pool (i.e., zone-1) there are the following files present:
> 
> * `<object>/xl.meta`
> * `<object>/0dcf7fc5-d7b6-4895-95f2-33da7c09189d/part.1`
> 
> So the object doesn't seem to be missing on any drive.

@haslersn we would love to see if you can share `xl.meta` from all disks and capture them as per order of the drives in your setup.

If you can share this we would love to take a look and see what might be causing this behavior. As this is unexpected and we have tried to reproduce we seem to be unable to.

**Comment by haslersn** [comment]
@harshavardhana Do the drives have an order? How to know the order? I'm using MinIO operator. Also, I redeployed the MinIO tenant once or twice, so the order might not have been preserved. The same drives were always plugged into the same servers.

**Comment by harshavardhana** [comment]
> @harshavardhana Do the drives have an order? How to know the order? I'm using MinIO operator. Also, I redeployed the MinIO tenant once or twice, so the order might not have been preserved. The same drives were always plugged into the same servers.

Orders if changed can have these consequences, did you remount the drives as well in different order ? If not it's not an issue.

**Comment by haslersn** [comment]
Each drives is mounted to a directory which is used as a local PersistentVolume (PV). So I don't see how mount order matters. But when creating the MinIO tenant, it will create so-called PersistentVolumeClaims (PVC) and the mapping between PVC and PV is determined essentially at random by Kubernetes.

But I certainly didn't mix drives from different pools, because I configured a different so-called StorageClass per pool. This also holds empirically: As I wrote earier, the files (`xl.meta` and `part.1`) belonging to said object can only be found on the 12 drives of zone-1.

I will upload the `xl.meta` files shortly.

**Comment by haslersn** [comment]
[minio-problem.txt](https://github.com/minio/minio/files/6872670/minio-problem.txt)

I couldn't upload with extension `.tar`, so I renamed it to `.txt`. It's a tar archive containing 12 files

```
<drive>-minio-ss-1-<server>.xl.meta
```

where `<server>` ranges from 0 to 2 and `<drive>` ranges from 0 to 3.

**Comment by harshavardhana** [comment]
> where `<server>` ranges from 0 to 2 and `<drive>` ranges from 0 to 3.

Can you provide the full command line? @haslersn

**Comment by harshavardhana** [comment]
@haslersn do you have compression enabled on the objects? in your setup?

**Comment by harshavardhana** [comment]
Your actual object version is perfectly fine 

- You seem to have objects compressed (i.e enabled server side compression)
- You seem to have enabled versioning on the buckets as well

```
~ ./xl-meta *.meta| jq .Versions[0].V2Obj.MTime
1622477080310506200
1622477080310506200
1622477080310506200
1622477080310506200
1622477080310506200
1622477080310506200
1622477080310506200
1622477080310506200
1622477080310506200
1622477080310506200
1622477080310506200
1622477080310506200

~ ./xl-meta *.meta| jq .Versions[0].V2Obj.ID
"fz6IJvHLSjuNepQvtTKFCQ=="
"fz6IJvHLSjuNepQvtTKFCQ=="
"fz6IJvHLSjuNepQvtTKFCQ=="
"fz6IJvHLSjuNepQvtTKFCQ=="
"fz6IJvHLSjuNepQvtTKFCQ=="
"fz6IJvHLSjuNepQvtTKFCQ=="
"fz6IJvHLSjuNepQvtTKFCQ=="
"fz6IJvHLSjuNepQvtTKFCQ=="
"fz6IJvHLSjuNepQvtTKFCQ=="
"fz6IJvHLSjuNepQvtTKFCQ=="
"fz6IJvHLSjuNepQvtTKFCQ=="
"fz6IJvHLSjuNepQvtTKFCQ=="
```

And this is in-tact

However you have inconsistent delete markers on the object i.e someone attempted a `DELETE` on the object of course since its versioned bucket it naturally creates a delete marker which is only present on 8 drives. 

```
~  ./xl-meta *.meta| jq .Versions[1].DelObj.ID
"beXjTzs5QqWfm0S41urLvA=="
"beXjTzs5QqWfm0S41urLvA=="
null
"beXjTzs5QqWfm0S41urLvA=="
"beXjTzs5QqWfm0S41urLvA=="
null
"beXjTzs5QqWfm0S41urLvA=="
"beXjTzs5QqWfm0S41urLvA=="
null
"beXjTzs5QqWfm0S41urLvA=="
"beXjTzs5QqWfm0S41urLvA=="
null
```

If you delete the delete marker or heal it - your object should still be accessible / readable IMO. This is not related to the same issue we are talking about in this issue.

**Comment by haslersn** [comment]
> Can you provide the full command line? @haslersn

What do you mean? I manually set the file names of the files in the tar, but they're based on the PVC names of the PVCs created by MinIO operator.

> @haslersn do you have compression enabled on the objects? in your setup?

I don't think so.

```shell
$ mc admin config get cluster01 compression
compression enable=off allow_encryption=off extensions=.txt,.log,.csv,.json,.tar,.xml,.bin mime_types=text/*,application/json,application/xml,binary/octet-stream 
```

Encryption is also turned off on the specific bucket.

> You seem to have enabled versioning on the buckets as well

Yes.

> If you delete the delete marker or heal it - your object should still be accessible / readable IMO.

`mc admin heal <object>` doesn't solve the problem. On the specific object the problem got, in the meantime, solved automatically (maybe by MinIO's automatic healing), but there are still other objects with the same problem. I just tried to heal one of them and the problem persists.

**Comment by haslersn** [comment]
I could solve the problem using `mc undo <object>`. Interestingly, this failed with an error

```
✓ Last delete of `data` is reverted.
mc: <ERROR> Unable to undo Resource requested is unreadable, please reduce your request rate
```

and the object was still deleted, but then calling `mc undo <object>` a second time worked without error. After that, reading the object was also possible. (Then I deleted it again, because the object was probably deleted on purpose).

I suspect the problem appears if an object gets deleted while one of the three servers is down (i.e., 4 of 12 drives down, per pool).

Why are you sure it's not related to the issue here? Should I create a separate issue (cause this should IMO be handled by MinIO without human intervention)?

**Comment by harshavardhana** [comment]
> I could solve the problem using `mc undo <object>`. Interestingly, this failed with an error
> 
> ```
> ✓ Last delete of `data` is reverted.
> mc: <ERROR> Unable to undo Resource requested is unreadable, please reduce your request rate
> ```
> 
> and the object was still deleted, but then calling `mc undo <object>` a second time worked without error. After that, reading the object was also possible. (Then I deleted it again, because the object was probably deleted on purpose).
> 
> I suspect the problem appears if an object gets deleted while one of the three servers is down (i.e., 4 of 12 drives down, per pool).
> 
> Why are you sure it's not related to the issue here? Should I create a separate issue (cause this should IMO be handled by MinIO without human intervention)?

Yes there is some weird problem here, this is related to something on delete markers and also delete markers written under quorum (i.e not all shards are written)

The failure is probably because of lack of quorum calculation on such objects and it's incorrect to return such errors. 

I will try to reproduce this with latest version of MinIO. 

I forgot to ask what is your server version?

**Comment by haslersn** [comment]
I upgraded from `RELEASE.2021-04-06T23-11-00Z` to `RELEASE.2021-06-17T00-10-46Z` to  `RELEASE.2021-07-15T22-27-34Z`.

**Comment by harshavardhana** [comment]
This also should be fixed with a whole bunch of fixes done here

- remove short-circuited healing optimization (#12796) (07/26/21) (Harshavardhana)
- fix: avoid removing healed parts on dstDataPath (#12795) (07/26/21) (Harshavardhana)
- improve delete-marker healing (#12794) (07/26/21) (Harshavardhana)
- add healing fixes for delete-marker (#12788) (07/26/21) (Harshavardhana)
